### ISSUE TITLE
Fastertransformer BERT returns wrong value in my environment.

### ISSUE THREAD

**Issue by ogis-uno** [triton-inference-server/fastertransformer_backend#60]
### Description

```shell
fastertransformer_backend: 225b57898b830
FasterTransformer: aa3aaf6d2ed0359ceb
Docker version: 20.10.17, build 100c701
NVIDIA Driver Version: 515.65.01
GPU: Tesla T4 x 1
```

Fastertransformer seems return wrong value for never seen input data in my environment.
(not always, but often)
Is there something wrong in my build and setup procedure?
I tried with NVIDIA Driver 460.32.03, 515.65.01, but no luck.

### Reproduced Steps
1. clone fastertransformer_backend repo.
```shell
git clone https://github.com/triton-inference-server/fastertransformer_backend.git
cd fastertransformer_backend
export WORKSPACE=$(pwd)
export CONTAINER_VERSION=22.07
export TRITON_DOCKER_IMAGE=triton_with_ft:${CONTAINER_VERSION}
```

2. Build container image.

```
docker build --rm   \
    --build-arg TRITON_VERSION=${CONTAINER_VERSION}   \
    -t ${TRITON_DOCKER_IMAGE} \
    -f docker/Dockerfile \
    .
```

3. Run container.

```
docker run -it --rm --gpus=all -v ${WORKSPACE}:${WORKSPACE} -w ${WORKSPACE} ${TRITON_DOCKER_IMAGE} bash
```

4. Convert Huggingface's BERT (in container)

```
export WORKSPACE=$(pwd)

sudo apt-get install git-lfs
git lfs install
git lfs clone https://huggingface.co/bert-base-uncased # Download model from huggingface
git clone https://github.com/NVIDIA/FasterTransformer.git # To convert checkpoint
export PYTHONPATH=${WORKSPACE}/FasterTransformer:${PYTHONPATH}
python3 FasterTransformer/examples/pytorch/bert/utils/huggingface_bert_convert.py \
        -in_file bert-base-uncased/ \
        -saved_dir ${WORKSPACE}/all_models/bert/fastertransformer/1/ \
        -infer_tensor_para_size 1
```

5. Modify config.pbtxt (due to `-infer_tensor_para_size 1`)

```
sed -i -e 's/string_value: "2"/string_value: "1"/' \
  -e "s#../all_models/bert/fastertransformer/1/2-gpu/#$WORKSPACE/all_models/bert/fastertransformer/1/1-gpu/#" all_models/bert/fastertransformer/config.pbtxt
rm -rf all_models/bert/fastertransformer/1/2-gpu/
```

6. Start Triton.

```
CUDA_VISIBLE_DEVICES=0,1 mpirun -n 1 --allow-run-as-root /opt/tritonserver/bin/tritonserver  --model-repository=${WORKSPACE}/all_models/bert/ &
```

7. Invoke test script (UserWarnings are removed for brevity).

```
root@77b40022d054:/home/uno/fastertransformer_backend# ./test.py --text "I have a pen."
Loading HF tokenizer and model...
Tokenizer text=I have a pen.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1024 06:52:51.197681 120 libfastertransformer.cc:1090] Start to forward
I1024 06:52:51.200114 120 libfastertransformer.cc:1098] Stop to forward
get results as output_hidden_state

ft_output_hidden_state: 
[[0. 0.]
 [0. 0.]
 [0. 0.]
 [0. 0.]
 [0. 0.]]
abs max diff: 8.78125
abs mean diff: 0.08430705964565277

root@77b40022d054:/home/uno/fastertransformer_backend# ./test.py --text "I have a pen."
Loading HF tokenizer and model...
Tokenizer text=I have a pen.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1024 06:53:26.158261 120 libfastertransformer.cc:1090] Start to forward
I1024 06:53:26.160828 120 libfastertransformer.cc:1098] Stop to forward
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.04846  0.01334]
 [ 0.227    0.4778 ]
 [ 0.3364   0.642  ]
 [ 0.1345   0.3196 ]
 [ 1.201    0.5034 ]]
abs max diff: 0.0390625
abs mean diff: 0.00032764553907327354

root@77b40022d054:/home/uno/fastertransformer_backend# ./test.py --text "I have an apple."
Loading HF tokenizer and model...
Tokenizer text=I have an apple.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1024 06:53:51.891521 120 libfastertransformer.cc:1090] Start to forward
I1024 06:53:51.894008 120 libfastertransformer.cc:1098] Stop to forward
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.04846  0.01334]
 [ 0.227    0.4778 ]
 [ 0.3364   0.642  ]
 [ 0.1345   0.3196 ]
 [ 1.201    0.5034 ]]
abs max diff: 2.24365234375
abs mean diff: 0.04817396402359009

root@77b40022d054:/home/uno/fastertransformer_backend# ./test.py --text "I have an apple."
Loading HF tokenizer and model...
Tokenizer text=I have an apple.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1024 06:54:18.815628 120 libfastertransformer.cc:1090] Start to forward
I1024 06:54:18.818453 120 libfastertransformer.cc:1098] Stop to forward
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.00159  0.181  ]
 [ 0.1421   0.4343 ]
 [ 0.2769   0.703  ]
 [-0.03543  1.12   ]
 [ 0.313    1.049  ]]
abs max diff: 0.015625
abs mean diff: 0.0003199076163582504
```

8. test.py

```
#!/usr/bin/python3
import argparse
import numpy as np
import sys
from builtins import range
import statistics as s
import tritonclient.grpc as grpcclient
import tritonclient.http as httpclient
from tritonclient.utils import np_to_triton_dtype
import random
import torch
from transformers import BertModel, AutoTokenizer


def prepare_tensor(name, input):
    t = grpcclient.InferInput(name, input.shape, np_to_triton_dtype(input.dtype))
    t.set_data_from_numpy(input)
    return t

def create_inference_server_client(protocol, url, concurrency, verbose):
    return grpcclient.InferenceServerClient(url, verbose=verbose)

def sequence_mask(lengths, max_len=None, is_2d=True):
    batch_size = lengths.numel()
    max_len = max_len or lengths.max()
    mask = (torch.arange(0, max_len, device=lengths.device)
            .type_as(lengths).repeat(batch_size, 1).lt(lengths.unsqueeze(1)))
    if is_2d:
        return mask
    else:
        mask = mask.view(-1, 1, 1, max_len)
        m2 = mask.transpose(2, 3)
        return mask * m2

def send_requests(url, input_hidden_state, sequence_lengths,
                  flags, request_parallelism=10):
    model_name = "fastertransformer"
    with create_inference_server_client(flags.protocol,
                                        url,
                                        concurrency=request_parallelism,
                                        verbose=False) as client:
        requests = []
        results = []
        
        for i in range(request_parallelism):
            inputs = [
                prepare_tensor("input_hidden_state", input_hidden_state),
                prepare_tensor("sequence_lengths", sequence_lengths),
            ]

            print("set request")
            result = client.infer(model_name, inputs)
            results.append(result)

        for i in range(request_parallelism):

            output_hidden_state = results[i].as_numpy("output_hidden_state")
            print("get results as output_hidden_state\n")
            if output_hidden_state is None:
                print("error: expected 'output_hidden_state'")
                sys.exit(1)
            else:
                pass
    return output_hidden_state

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("--text", type=str, default="I have an apple.", help="text data")
    FLAGS = parser.parse_args()
    FLAGS.protocol = "grpc"
    FLAGS.url = "localhost:8001"
    FLAGS.seq_len=32
    FLAGS.hf_ckpt_path="bert-base-uncased"
    FLAGS.inference_data_type="fp16"

    torch.manual_seed(0)
    random.seed(0)
    np.random.seed(0)

    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

    print("Loading HF tokenizer and model...")
    tokenizer = AutoTokenizer.from_pretrained(FLAGS.hf_ckpt_path)
    hf_model = BertModel.from_pretrained(FLAGS.hf_ckpt_path)
    hf_model.cuda().eval().to(torch.float16)

    print("Tokenizer text={}".format(FLAGS.text))
    encoded = tokenizer.batch_encode_plus([FLAGS.text], return_tensors="pt", padding="max_length", max_length=FLAGS.seq_len)
    encoded = {k:v.to(device) for k, v in encoded.items()}

    print("Prepare input hidden state")
    input_hidden_state = hf_model.embeddings(input_ids=encoded["input_ids"], token_type_ids=encoded["token_type_ids"])
    input_hidden_state = input_hidden_state.cpu().detach().numpy().astype(np.float16)
    sequence_lengths = torch.sum(encoded['attention_mask'], axis=1).cpu().detach().numpy()[np.newaxis,:].astype(np.int32)

    mask = sequence_mask(torch.Tensor(sequence_lengths.reshape([-1])).to(torch.int), FLAGS.seq_len, False)
    mask = mask.to(torch.float16).cuda()
    output_mask = sequence_mask(torch.Tensor(sequence_lengths.reshape([-1])).to(torch.int), FLAGS.seq_len)
    output_mask = output_mask.to(mask.dtype).unsqueeze(-1).cuda()
    
    print("Invoke Huggingface")
    hf_input = torch.Tensor(input_hidden_state).cuda().to(torch.float16)
    extended_attention_mask = (1.0 - mask) * -10000.0
    hf_output = hf_model.encoder(hf_input, extended_attention_mask, return_dict=False)[0] * output_mask

    print("Invoke FasterTransformer")
    ft_output_hidden_state = send_requests(FLAGS.url, input_hidden_state, sequence_lengths,
                                           FLAGS, request_parallelism=1)

    print("ft_output_hidden_state: \n{}".format(ft_output_hidden_state[0, :5, :2]))
    diff = (torch.Tensor(ft_output_hidden_state).cuda() * output_mask) - hf_output

    print(f"abs max diff: {diff.abs().max()}")
    print(f"abs mean diff: {diff.abs().mean()}")
```
```

**Comment by ogis-uno** [comment]
Hi byshiue,  Thanks for reply!

But I'm not familiar with cmake...  The following procedure is correct?

1. Update Bert.cc
```
$ sudo sed -i '681a\ \ cudaStreamSynchronize(stream_);' FasterTransformer/src/fastertransformer/models/bert/Bert.cc
$ cd FasterTransformer && git diff
diff --git a/src/fastertransformer/models/bert/Bert.cc b/src/fastertransformer/models/bert/Bert.cc
index 7acbb97..82aa84e 100644
--- a/src/fastertransformer/models/bert/Bert.cc
+++ b/src/fastertransformer/models/bert/Bert.cc
@@ -679,6 +679,7 @@ void Bert<T>::forward(TensorMap* output_tensors, TensorMap* input_tensors, const
         // throw errors when detected
         ftNcclStreamSynchronize(tensor_para_, pipeline_para_, stream_);
     }
+  cudaStreamSynchronize(stream_);
 }
 
 template class Bert<float>;
```

2. Run container.

```
$ docker run -it -v ${WORKSPACE}:/workspace --name ft_backend_builder ${TRITON_DOCKER_IMAGE} bash
```

3. Rebuild . ( in container)
```
# mkdir build -p && cd build && \
  cmake -D CMAKE_EXPORT_COMPILE_COMMANDS=1  \
             -D CMAKE_BUILD_TYPE=Release  \
             -D CMAKE_INSTALL_PREFIX=/opt/tritonserver  \
             -D TRITON_COMMON_REPO_TAG="r${NVIDIA_TRITON_SERVER_VERSION}"  \
             -D TRITON_CORE_REPO_TAG="r${NVIDIA_TRITON_SERVER_VERSION}"  \
             -D TRITON_BACKEND_REPO_TAG="r${NVIDIA_TRITON_SERVER_VERSION}"  .. &&  \
   make -j"$(grep -c ^processor /proc/cpuinfo)" install
```

4. Commit change.

```
$ docker commit ft_backend_builder ${TRITON_DOCKER_IMAGE}
```

Or, Shoud I fork FastarTransformer repo and modify it,  and update GIT_REPOSITORY/GIT_TAG in CMakeLists.txt in fastertransformer_backend?

**Comment by byshiue** [comment]
Or let's use another simpler way to check first. 

Can you try to wait some seconds after each inference, and then receive it?

**Comment by ogis-uno** [comment]
> Can you try to wait some seconds after each inference, and then receive it?

You mean like this ?
```
# modified send_requests() of my test.py
...
        for i in range(request_parallelism):
            inputs = [
                prepare_tensor("input_hidden_state", input_hidden_state),
                prepare_tensor("sequence_lengths", sequence_lengths),
            ]

            print("set request")
            result = client.infer(model_name, inputs)
            import time, datetime
            print("Start sleep " + datetime.datetime.now().strftime('%H:%M:%S'))
            time.sleep(5)
            print("End sleep " + datetime.datetime.now().strftime('%H:%M:%S'))

            results.append(result)

        for i in range(request_parallelism):

            output_hidden_state = results[i].as_numpy("output_hidden_state")
...
```

I tried with modification above, but got no lock.
```
root@54ae478b14b8:/home/uno/fastertransformer_backend# ./test.py --text "I have a pen."
Loading HF tokenizer and model...
Tokenizer text=I have a pen.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1024 09:11:20.430276 95 libfastertransformer.cc:1090] Start to forward
I1024 09:11:20.432580 95 libfastertransformer.cc:1098] Stop to forward
Start sleep 09:11:20
End sleep 09:11:25
get results as output_hidden_state

ft_output_hidden_state: 
[[0. 0.]
 [0. 0.]
 [0. 0.]
 [0. 0.]
 [0. 0.]]
abs max diff: 8.78125
abs mean diff: 0.08430705964565277
root@54ae478b14b8:/home/uno/fastertransformer_backend# ./test.py --text "I have a pen."
Loading HF tokenizer and model...
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1024 09:11:59.219416 95 libfastertransformer.cc:1090] Start to forward
I1024 09:11:59.221243 95 libfastertransformer.cc:1098] Stop to forward
Start sleep 09:11:59
End sleep 09:12:04
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.04846  0.01334]
 [ 0.227    0.4778 ]
 [ 0.3364   0.642  ]
 [ 0.1345   0.3196 ]
 [ 1.201    0.5034 ]]
abs max diff: 0.0390625
abs mean diff: 0.00032764553907327354
root@54ae478b14b8:/home/uno/fastertransformer_backend# ./test.py --text "I have an apple."
Loading HF tokenizer and model...
Tokenizer text=I have an apple.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1024 09:12:24.021994 95 libfastertransformer.cc:1090] Start to forward
I1024 09:12:24.024597 95 libfastertransformer.cc:1098] Stop to forward
Start sleep 09:12:24
End sleep 09:12:29
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.04846  0.01334]
 [ 0.227    0.4778 ]
 [ 0.3364   0.642  ]
 [ 0.1345   0.3196 ]
 [ 1.201    0.5034 ]]
abs max diff: 2.24365234375
abs mean diff: 0.04817396402359009
root@54ae478b14b8:/home/uno/fastertransformer_backend# ./test.py --text "I have an apple."
Loading HF tokenizer and model...
Tokenizer text=I have an apple.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1024 09:12:46.888630 95 libfastertransformer.cc:1090] Start to forward
I1024 09:12:46.890466 95 libfastertransformer.cc:1098] Stop to forward
Start sleep 09:12:46
End sleep 09:12:51
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.00159  0.181  ]
 [ 0.1421   0.4343 ]
 [ 0.2769   0.703  ]
 [-0.03543  1.12   ]
 [ 0.313    1.049  ]]
abs max diff: 0.015625
abs mean diff: 0.0003199076163582504
```

**Comment by byshiue** [comment]
I try to fix it in `main` branch. Can you re-build the docker and test again?

**Comment by byshiue** [comment]
test

**Comment by ogis-uno** [comment]
> Can you re-build the docker and test again?

Yes, absolutely!  Please let me know when you fixed.

**Comment by ogis-uno** [comment]
I found your fix f59e237c24 in FasterTransformer repo!
So I rebuild docker image, and test again.

In container...
I confirmed your fix applied.
```
root@b354fad7591e:/home/uno/fastertransformer_backend# grep cudaStreamSynchronize /workspace/build/fastertransformer_backend/build/_deps/repo-ft-src/src/fastertransformer/models/bert/Bert.cc 
    cudaStreamSynchronize(stream_);
```
Then test again. 
```
root@b354fad7591e:/home/uno/fastertransformer_backend# ./test.py --text "I have a pen."
Loading HF tokenizer and model...
Tokenizer text=I have a pen.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1025 06:08:01.144699 713 libfastertransformer.cc:1090] Start to forward
I1025 06:08:01.148176 713 libfastertransformer.cc:1098] Stop to forward
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.04846  0.01334]
 [ 0.227    0.4778 ]
 [ 0.3364   0.642  ]
 [ 0.1345   0.3196 ]
 [ 1.201    0.5034 ]]
abs max diff: 0.0390625
abs mean diff: 0.00032764553907327354

root@b354fad7591e:/home/uno/fastertransformer_backend# ./test.py --text "I have a pen."
Loading HF tokenizer and model...
Tokenizer text=I have a pen.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1025 06:08:11.796973 713 libfastertransformer.cc:1090] Start to forward
I1025 06:08:11.799804 713 libfastertransformer.cc:1098] Stop to forward
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.04846  0.01334]
 [ 0.227    0.4778 ]
 [ 0.3364   0.642  ]
 [ 0.1345   0.3196 ]
 [ 1.201    0.5034 ]]
abs max diff: 0.0390625
abs mean diff: 0.00032764553907327354


root@b354fad7591e:/home/uno/fastertransformer_backend# ./test.py --text "I have an apple."
Loading HF tokenizer and model...
Tokenizer text=I have an apple.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1025 06:08:26.260781 713 libfastertransformer.cc:1090] Start to forward
I1025 06:08:26.263680 713 libfastertransformer.cc:1098] Stop to forward
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.00159  0.181  ]
 [ 0.1421   0.4343 ]
 [ 0.2769   0.703  ]
 [-0.03543  1.12   ]
 [ 0.313    1.049  ]]
abs max diff: 0.015625
abs mean diff: 0.0003199076163582504


root@b354fad7591e:/home/uno/fastertransformer_backend# ./test.py --text "I have an apple."
Loading HF tokenizer and model...
Tokenizer text=I have an apple.
Prepare input hidden state
Invoke Huggingface
Invoke FasterTransformer
set request
I1025 06:08:36.563991 713 libfastertransformer.cc:1090] Start to forward
I1025 06:08:36.567049 713 libfastertransformer.cc:1098] Stop to forward
get results as output_hidden_state

ft_output_hidden_state: 
[[-0.00159  0.181  ]
 [ 0.1421   0.4343 ]
 [ 0.2769   0.703  ]
 [-0.03543  1.12   ]
 [ 0.313    1.049  ]]
abs max diff: 0.015625
abs mean diff: 0.0003199076163582504
```
I got same data! 
I tried few times more,  and got same result every time. I think it works fine!
Thank you.
